Provide consistent exit codes

bertschneider opened the pull request:
This change uses the possibility to stop the fuzzer loop provided in the patched
libFuzzer version. Incorporating this allows the removal of the stopFuzzing
functions and general simplifications of the fuzzing process.

The core fuzzing function now returns the expected exit code and found error so
that invoking code can handle them according to its needs, e.g. exit the
process with the given code in case of CLI, and rethrow the error in case of
Jest.

oetr reviewed (commented):
Thanks for this awesome PR! It makes a large chunk of our native code much clearer.

bertschneider commented on packages/core/finding.ts:
```diff
@@ -42,25 +56,37 @@ export function clearFirstFinding(): Finding | undefined {
+ * @param findingMessage - The finding to be saved.
+ * @param containStack - Whether the finding should contain a stack trace or not.
+ */
+export function reportFinding(
```
Oh, yes, that's way better! Changed!

bertschneider commented on packages/core/core.ts:
```diff
@@ -152,150 +162,166 @@ function getFilteredBugDetectorPaths(
-		// primitive types
-		return String(error);
+// Wrap fuzz target to print and dump the crashing input on error.
+function asCrashDumpFuzzFn(fuzzFn: fuzzer.FuzzTarget) {
```
Exactly, the additional call from `jest-runner` is the problem. I tried to come up with a nice solution but failed. I will give it another try, though, as I also would like to have only one wrapper around the fuzz target.

bertschneider replied:
Pushed another version without the additional wrapper and enforced types.

bertschneider commented on packages/core/core.ts:
```diff
@@ -152,150 +162,166 @@ function getFilteredBugDetectorPaths(
-			);
-		},
-	);
+	const findingAwareFuzzFn = asFindingAwareFuzzFn(fuzzFn);
```
Actually, it was there before :see_no_evil: My reasoning on extracting the call was that `loadFuzzFunction` did two things, as the name suggests loading the actual fuzz function, but also wrapping that to report findings. I find it a little bit cleaner and easier to follow this way. We could change it back, though, no hard feeling either way.

bertschneider commented on .gitignore:
```diff
@@ -53,4 +54,5 @@ node_modules/
 # Dictionaries generated by Jazzer.js
-.JazzerJs-merged-dictionaries
\ No newline at end of file
+.JazzerJs-merged-dictionaries
```
Good catch! Deleted.

bertschneider commented:
I incorporated the latest signal handling changes into this PR. It seems that handling SIGSEGV in the native addon only is sufficient. SIGINT still works without a handler in node, so we can get rid of all JS signal handling code.
Furthermore, I changed the sigsegv error message so that it's recognized by cifuzz. 

This leaves Jazzer.js with consistent error messages, except for timeouts, where libFuzzer kills the process with a timeout crash and exit code `70`, and segfaults, where the addon kills the process with exit code `139`.
